### PR TITLE
Update hackage-security to GHC 8.8

### PR DESCRIPTION
**Pull request by Avi-D-coder**
This goes on top of #232

### CODE REVIEW

**Comment by Avi-D-coder** [comment]
Reworking this

**Comment by Avi-D-coder** [comment]
Is `MonadFail` the right approach?

File: hackage-security/src/Hackage/Security/Client.hs [hunk 1]
```diff
@@ -335,6 +338,8 @@ cachedVersion CachedInfo{..} remoteFile =
 getCachedInfo ::
 #if __GLASGOW_HASKELL__ < 800
                  (Applicative m, MonadIO m)
+#elif __GLASGOW_HASKELL__ >= 807
+                 (MonadFail ReadJSON_Keys_Layout, MonadIO m)
```
Comment (bgamari) [inline hackage-security/src/Hackage/Security/Client.hs]: Indeed this seems rather odd. I haven't looked at the surrounding code but it does seem suspicious that you would need to add a constraint of this form. Are you sure this is necessary?

File: hackage-security/src/Hackage/Security/TUF/Snapshot.hs [hunk 2]
```diff
@@ -68,6 +68,7 @@ instance MonadReader RepoLayout m => ToJSON m Snapshot where
 
 instance ( MonadReader RepoLayout m
          , MonadError DeserializationError m
+         , MonadFail m
```
Comment (bgamari) [inline hackage-security/src/Hackage/Security/TUF/Snapshot.hs]: In this case I would probably try to use the `MonadError` constraint to instead report a more precise error. For instance, define locally:
```haskell
let lookupFile path =
      withExceptT DeserializationErrorSchema 
      $ liftEither $ FileMap.lookupM snapshotMeta path
```
and use this in place of `FIleMap.lookupM` below.

File: hackage-security/src/Hackage/Security/Client.hs (hunk 1 above)
Comment (Avi-D-coder) [reply hackage-security/src/Hackage/Security/Client.hs]: I'm pretty sure I did it all wrong, I'll dig into it tomorrow.

File: hackage-security/src/Hackage/Security/TUF/Snapshot.hs (hunk 2 above)
Comment (Avi-D-coder) [reply hackage-security/src/Hackage/Security/TUF/Snapshot.hs]: I was not able to make this work so I just wrote:
```haskell
lookupErr :: MonadError e m => (String -> e) -> FileMap -> TargetPath -> m FileInfo
```

File: hackage-security/src/Hackage/Security/Client.hs (hunk 1 above)
Comment (Avi-D-coder) [reply hackage-security/src/Hackage/Security/Client.hs]: I'm not clear how `MonadFail` can be replaced, it seems to come from `getCachedInfo`:`347` and `ReadJSON_Keys_Layout` comes from  `readLocalFile `: `367`.

Should I try to replace `MonadFail` with a `Throws _`?

Comment (Avi-D-coder) [reply hackage-security/src/Hackage/Security/Client.hs]: My `mtl` foo is not good enough to remove `MonadFail ReadJSON_Keys_Layout`. 

@edsko I see you wrote the `ReadJSON_Keys_Layout` code, do you have any advice?

Comment (Avi-D-coder) [reply hackage-security/src/Hackage/Security/Client.hs]: THe cloest I have gotten is `MonadError DeserializationError IO`, but this produces the warning:

```haskell
Security/Client.hs:211:18: warning:
    • Couldn't match type ‘IOException’ with ‘DeserializationError’
        arising from a functional dependency between:
          constraint ‘MonadError DeserializationError IO’
            arising from the type signature for:
                           checkForUpdates :: forall (down :: * -> *).
                                              (MonadError DeserializationError IO,
                                               Throws VerificationError, Throws SomeRemoteError) =>
                                              Repository down -> Maybe UTCTime -> IO HasUpdates
          instance ‘MonadError IOException IO’ at <no location info>
      Inaccessible code in
        a pattern with constructor: FGz :: Format FormatGz,
        in a case alternative
    • In the pattern: FGz
      In the pattern: Some FGz
      In a case alternative:
          Some FGz -> verifyFileInfo' (Just newInfoTarGz) targetPath tempPath
```

**Comment by hvr** [comment]
@Avi-D-coder to be honest, I'm utterly confused by this PR; is this really against master? Also this seems to address an issue that has long been resolved by 471a7b8ec1867cc4b641b7846912252e74637b87, no?

File: hackage-security/hackage-security.cabal [hunk 3]
```diff
@@ -99,8 +99,8 @@ library
                        base16-bytestring >= 0.1.1   && < 0.2,
                        base64-bytestring >= 1.0     && < 1.1,
                        bytestring        >= 0.9     && < 0.11,
-                       Cabal             >= 1.14    && < 1.26,
```
Comment (hvr) [inline hackage-security/hackage-security.cabal]: This line here makes me suspect this is against some fairly ancient state of the master branch

**Comment by bgamari** [comment]
Hmm, indeed I am also a bit confused; `master` already appears to build with GHC 8.8.1. @Avi-D-coder, is there a particular error that you were seeing?

**Comment by Avi-D-coder** [comment]
This does appear to be against an older master

**Pull request closed without merging**